highfidelity/hifi
# Update to Transitions debugging / editing script

This PR adds some small improvements to the debugTransition.js script in _scripts\developer\utilities\render_. This script can be used to visualize the various transition effects in their current state (some of which are not yet plugged) and tweak some of their parameters. Here are the introduced modifications:

- Load / Save can now load and save a transition effect configuration from any JSON file with the help of classic file dialog. The file dialog will only work in desktop mode.
- Each transition event, when selected, displays a small textual description of its behaviour.
- Transition events that are not time based will have the timing parameter widgets hidden.
- Transition edit mode is active when the transition edit app is opened and inactive when the app is hidden.
- Minor UI layout changes.

![transitioneditor](https://user-images.githubusercontent.com/14199381/38150764-fa182f28-3436-11e8-9239-ecd3782352b0.jpg)

The Transition edit item selection method is now based on ray picking. It works with the help of a new selection called "TransitionEdit" which is used to add items that should have the currently edited transition effect (with the help of the debugTransition.js script) applied to them. 

# Test Plan
This test should be performed in both desktop and HMD mode.

1. Run the _scripts\developer\utilities\render\debugTransition.js_ script : this should add a new tablet app named "Transition"
2. Open the app and pick a 3D object with the mouse : the "Element Enter/Leave" transition effect should be applied repeatedly on that object.
3. Close the app : the transition effect should disappear
4. Re-open the app : the transition effect should reappear on the same object
5. Clicking on another object should apply the transition effect to that new object while removing the effect from the previous.
6. Modifying any of the transition effect parameters should have a visual impact
7. Changing transition event types should modify the textual description
8. When any of the "Bubble Isect..." events are selected, the duration slider and timing box should both be invisible.
9. Clicking on the "Save" button should open a file dialog and saving to a file should produce a non empty JSON file
10. Clicking on the "Load" button should open a file dialog and opening a previously saved JSON file should apply the settings to the current transition effect.
